autoscheduler/scraper/banner_requests.py

```diff
@@ -149,24 +148,58 @@ def get_departments(self, amount: int = 300) -> List[Dict]:
 
         return depts
 
-    async def search(self, depts: List[str], amount: int = 750) -> List[List[Dict]]:
+    async def search(self, depts_terms: List[Tuple[str, str]],
+                     sem: asyncio.Semaphore,
+                     parse_all_courses: Callable[[list], list],
+                     amount: int = 750
+                    ) -> List[List[Dict]]:
         """ Concurrently retrieves all of the given departments and returns them as
             a list of course-lists, with each index corresponding to the courses/sections
             for a department.
         """
 
         loop = asyncio.get_running_loop()
 
-        async def perform_search(dept: str):
-            result = None
+        courses_set = set()
+        instructors_set = set()
+
+        async def perform_search(dept: str, term: str):
+            result = []
+
             async with ClientSession(loop=loop) as session:
-                session_id = await self.create_session(session)
+                course_list = []
+
+                retry_max = 10
+                for i in range(1, retry_max + 1):
+                    try:
+                        # Must limit the amoutn of concurrent requests, otherwise will get
+                        # a "too many file descriptors in select()" error from aiohttp
+                        async with sem:
+                            print(f"Starting {dept} {term}")
+                            session_id = await self.create_session(session, term)
+
+                            course_list = await self.get_courses(session, session_id,
+                                                                 dept, term, amount)
+
+                        # We only want to limit the requests, not parsing, so call this
+                        # outside of the semaphore
+                        ret = parse_all_courses(course_list, term, courses_set,
+                                                instructors_set)
+
+                        if ret is False:
+                            continue # Failure (course_list was none), retry
+
+                        result.extend(ret)
 
-                result = await self.get_courses(session, session_id, dept, amount)
+                    except (ClientConnectorError, ContentTypeError):
+                        # Empty lines help it stand out from the rest of the outputs
+                        print(f"\n\nNETWORK ERROR: Retrying {dept} {term}: Take {i} \n\n")
+                        continue # Error occurred, retry
+                    break # If we pass it, no need to keep looping

```

Comment:
Move the `break` up to the end of the `try` block and remove the `continue`, I was confused looking at this at first because I didn't realize the `break` was supposed to execute after the `try` block finishes

Reply:
Yeah this makes much more sense, will fix